consensus/src/network_interface.rs

```diff
@@ -117,18 +120,49 @@ impl NewNetworkSender for ConsensusNetworkSender {
 }
 
 impl ConsensusNetworkSender {
+    /// Initialize a shared hashmap about connections metadata that is updated by the receiver.
+    pub fn initialize(&mut self, connections: Arc<RwLock<HashMap<PeerId, ProtocolIdSet>>>) {
+        self.peers_protocols = Some(connections);
+    }
+
+    /// Query the supported protocols from this peer's connection.
+    fn supported_protocols(&self, peer: PeerId) -> anyhow::Result<ProtocolIdSet> {
+        if let Some(map) = &self.peers_protocols {
+            map.read()
+                .get(&peer)
+                .cloned()
+                .ok_or_else(|| anyhow!("Peer not connected"))
+        } else {
+            Err(anyhow!("ConsensusNetworkSender not initialized"))
+        }
+    }
+
+    /// Choose the overlapping protocol for peer. The local protocols are sorted from most to least preferred.
+    fn preferred_protocol_for_peer(
+        &self,
+        peer: PeerId,
+        local_protocols: &[ProtocolId],
+    ) -> anyhow::Result<ProtocolId> {
+        let remote_protocols = self.supported_protocols(peer)?;
+        for protocol in local_protocols {
+            if remote_protocols.contains(*protocol) {
+                return Ok(*protocol);
+            }
+        }
```

Comment:
I bet you could find a one liner for this, something like

local_protocols.rfind(|&protocol| remote_protocols.contains(protocol)).ok_or_err(...)

my rust is rusty...

Reply:
I will follow up on refactoring this afterwards.  I didn't write it, and I only moved it's location to homogenize the network senders :P